stellargraph/layer/watch_your_step.py

```diff
@@ -0,0 +1,199 @@
+import tensorflow as tf
+from tensorflow.keras.layers import Layer, Embedding, Input, Lambda, Concatenate
+from tensorflow.keras import backend as K
+from tensorflow.keras import regularizers, initializers, constraints
+import numpy as np
+
+from ..mapper.adjacency_generators import AdjacencyPowerGenerator
+from ..core.experimental import experimental
+
+
+@experimental(reason="lack of unit tests")
+class AttentiveWalk(Layer):
+    """
+    This implements the graph attention as in Watch Your Step: Learning Node Embeddings via Graph Attention
+    https://arxiv.org/pdf/1710.09599.pdf.
+
+    .. warning::
+
+        This class is experimental: it is insufficiently tested.
+
+    Args:
+        walk_length (int): the length of the random walks. Equivalent to the number of adjacency powers used.
+        attention_initializer (str or func): The initialiser to use for the attention weights;
+            defaults to 'glorot_uniform'.
+        attention_regularizer (str or func): The regulariser to use for the attention weights;
+            defaults to None.
+        attention_constraint (str or func): The constraint to use for the attention weights;
+            defaults to None.
+    """
+
+    def __init__(self, walk_length=5, **kwargs):
+        if "input_shape" not in kwargs and "input_dim" in kwargs:
+            kwargs["input_shape"] = (kwargs.get("input_dim"),)
+
+        self.walk_length = walk_length
+
+        self._get_regularisers_from_keywords(kwargs)
+        super().__init__(**kwargs)
+
+    def compute_output_shape(self, input_shapes):
+        """
+        Computes the output shape of the layer.
+        Assumes the following inputs:
+
+        Args:
+            input_shapes (tuple of ints)
+                Shape tuples can include None for free dimensions, instead of an integer.
+
+        Returns:
+            An input shape tuple.
+        """
+        return (input_shapes[0][-1],)
+
+    def _get_regularisers_from_keywords(self, kwargs):
+        self.attention_initializer = initializers.get(
+            kwargs.pop("attention_initializer", "glorot_uniform")
```

Comment:
I was just keeping things consistent with some of the other `stellargraph` stuff - but I think real arguments make more sense here. I'll keep "input_shape"/"input_dim" in the `kwargs` for now just so someone isn't tripped up by having to specify "input_shape" now when previously "input_dim" worked fine for other `stellargraph` layers.

Reply:
> I was just keeping things consistent with some of the other stellargraph stuff - but I think real arguments make more sense here

Yeah, I don't blame you: consistency is good and important. However, in this case, I think we should be moving everything to arguments where possible, especially because this doesn't change how the user interacts with the API, just how it is documented. 👍 

> I'll keep "input_shape"/"input_dim" in the kwargs for now just so someone isn't tripped up by having to specify "input_shape" now when previously "input_dim" worked fine for other stellargraph layers.

I wasn't suggesting changing the functionality/the API, just how it is expressed, similar to how the `attention_initializer="glorot_uniform"` defaulted-parameter works the same as the manual defaults done by manually examining `kwargs`; I could be wrong, but I think the current `input_dim` behaviour is equivalent to:

```python
def __init__(self, ..., input_dim=None, **kwargs):
    if "input_shape" not in kwargs and input_dim is not None:
        kwargs["input_dim"] = (input_dim,)
```

Reply:
I've included `input_dim` as a real argument now :)